### PR TITLE
[Coral-762] Support multiple byte streams for one block transfer

### LINKED ISSUE #762: Support multiple byte streams per one block request
When we retrieve compressed multiple partitions, we should know where is the each partition's boundary to decompress each partition completely. If inflating stream reads concatenated stream, it will read only the first partition and the rests will be missed. So we need to create multiple byte streams on one block request to handle such situation.

### PR DESCRIPTION
**Pull request by seojangho**
This commit features the following changes:

* Add support for multiple sub-stream for one block transfer context
* Complete separation of block management and byte transfer
  * Lower data plane renamed from 'BlockTransfer' to 'ByteTransfer'
  * Higher data plane embeds block metadata in 'TransferContext'.
    Lower data plane does not interpret it.
* Enabling Streaming for inter-TaskGroup communication (disabled by BlockManagerWorker)
* Fixed InputReader not to introduce deadlock while reading streamized SideInput
* Much more concise codebase!

Resolves #762

### CODE REVIEW

**Comment by seojangho** [comment]
@sanha I've address the issues we discussed offline. Please have a look.

File: runtime/executor/src/main/java/edu/snu/coral/runtime/executor/bytetransfer/ContextManager.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+final class ContextManager extends SimpleChannelInboundHandler<ByteTransferContextSetupMessage> {
+
+  private final BlockManagerWorker blockManagerWorker;
+  private final ByteTransfer byteTransfer;
+  private final ChannelGroup channelGroup;
+  private final String localExecutorId;
+  private final Channel channel;
+  private volatile String remoteExecutorId = null;
+
+  private final ConcurrentMap<Integer, ByteInputContext> inputContextsInitiatedByLocal = new ConcurrentHashMap<>();
```
Comment (sanha) [inline runtime/executor/src/main/java/edu/snu/coral/runtime/executor/bytetransfer/ContextManager.java]: Why do we use `Map` between index and object? (Why not `List`?)

File: runtime/executor/src/main/java/edu/snu/coral/runtime/executor/data/BlockManagerWorker.java [hunk 2]
```diff
@@ -371,41 +388,66 @@ public void run() {
...
    * thread. Never execute a blocking call in this method!
    *
-   * @param inputStream {@link BlockInputStream}
+   * @param inputContext {@link ByteInputContext}
    */
-  public void onPushNotification(final BlockInputStream inputStream) {
+  public void onInputContext(final ByteInputContext inputContext) {
+  }
+
+  private static ControlMessage.BlockStore convertBlockStore(
```
Comment (sanha) [inline runtime/executor/src/main/java/edu/snu/coral/runtime/executor/data/BlockManagerWorker.java]: Please add simple comments for these methods.

File: runtime/executor/src/main/java/edu/snu/coral/runtime/executor/data/blocktransfer/BlockInputStream.java [hunk 3]
```diff
@@ -1,288 +0,0 @@
...
-  }
-
-  /**
-   * Gets a {@link CompletableFuture} that completes with the block transfer being done.
-   * This future is completed by one of the decoding thread. Consider using separate {@link ExecutorService} when
-   * chaining a task to this future.
-   *
-   * @return a {@link CompletableFuture} that completes with the block transfer being done
-   */
-  public CompletableFuture<Iterator<T>> getCompleteFuture() {
```
[the remark this answers is not in the document]
Comment (sanha) [reply runtime/executor/src/main/java/edu/snu/coral/runtime/executor/data/blocktransfer/BlockInputStream.java]: Oh, you addressed it already! Never mind.

File: runtime/executor/src/main/java/edu/snu/coral/runtime/executor/bytetransfer/ContextManager.java (hunk 1 above)
Comment (seojangho) [reply runtime/executor/src/main/java/edu/snu/coral/runtime/executor/bytetransfer/ContextManager.java]: Two reasons:
- To detect (potential) transferIndex conflict
- Used by `ContextManager#getInputContext` method to provide incoming data on the corresponding context.

File: runtime/executor/src/main/java/edu/snu/coral/runtime/executor/bytetransfer/ByteTransport.java [hunk 4]
```diff
@@ -193,43 +188,62 @@ public void close() {
...
+      if (future.isSuccess()) {
+        // Succeed to connect
+        LOG.debug("Connected to {}", remoteExecutorId);
+        return;
+      }
+      // Failed to connect
+      if (future.cause() == null) {
+        LOG.error("Failed to connect to {}", remoteExecutorId);
+      } else {
+        LOG.error(String.format("Failed to connect to %s", remoteExecutorId), future.cause());
```
[the remark this answers is not in the document]
Comment (seojangho) [reply runtime/executor/src/main/java/edu/snu/coral/runtime/executor/bytetransfer/ByteTransport.java]: Yes. See [ByteTransfer#conectTo](https://github.com/snuspl/coral/pull/765/files/1e0c49d88cb476d14dc47b020abdf3a0aa78f617#diff-33252990958fc9f4e5789ee12520e1d9R97).

**Comment by seojangho** [comment]
@sanha Thanks for the comments. Please check the updates.

**Comment by seojangho** [comment]
!rebuild

**Comment by seojangho** [comment]
!rebuild

**Pull request merged**